### PR TITLE
Update to GraphHopper ors_v0.13.8

### PR DESCRIPTION
**Pull request by aoles**
### Pull Request Checklist

- [x] 1. I have [**rebased**](https://github.com/GIScience/openrouteservice/blob/master/CONTRIBUTE.md#pull-request-guidelines) the latest version of the master branch into my feature branch and all conflicts have been resolved.
- [x] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the [Unreleased] heading.
- [ ] 3. I have documented my code using JDocs tags.
- [x] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [ ] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [ ] 6. I have created API tests for any new functionality exposed to the API.
- [ ] 7. If changes/additions are made to the app.config file, I have added these to the app.config wiki page on github along with a short description of what it is for, and documented this in the Pull Request (below).
- [x] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [ ] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [x] 10. For new features or changes involving building of graphs, I have tested on a larger dataset (at least Germany) and the graphs build without problems (i.e. no out-of-memory errors).
- [x] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the importer etc.), I have generated longer distance routes for the affected profiles with different options (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end points generated from the current live ORS. If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [x] 12. I have written in the Pull Request information about the changes made including their intended usage and why the change was needed.

### Information about the changes
- Key functionality added:
Introduce time-dependent Dijkstra and A* into ORS.
- Reason for change:
Base for for time-dependent isochrone or route calculation.

### CODE REVIEW

**Comment by HendrikLeuschner** [comment]
Why are the param tests getting a connection refused error?

**Comment by takb** [comment]
my bad, messed up the gh release (tagged the wrong branch...), I'm working on fixing it.

File: openrouteservice/src/main/java/org/heigit/ors/routing/RoutingProfile.java [hunk 1]
```diff
@@ -989,7 +1003,8 @@ private void setWeighting(HintsMap map, int requestWeighting, int profileType){
         //For a requested recommended weighting, use recommended for bike, walking and hgv. Use fastest for car.
         if (requestWeighting == WeightingMethod.RECOMMENDED || requestWeighting == WeightingMethod.FASTEST) {
             if (profileType == RoutingProfileType.DRIVING_CAR) {
-                weighting = weightingMethod = VAL_FASTEST;
+                weighting = VAL_FASTEST;
```
Comment (HendrikLeuschner) [inline openrouteservice/src/main/java/org/heigit/ors/routing/RoutingProfile.java:1009]: Do we only support car with td? Also can we make "td_fastest" a String variable like the others?

File: openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java [hunk 2]
```diff
@@ -361,6 +361,29 @@ else if (ALT_ROUTE.equalsIgnoreCase(algoStr))
 				if (weighting instanceof TurnWeighting)
 	                ((TurnWeighting)weighting).setInORS(true);
 
+				weighting = createTimeDependentAccessWeighting(weighting, tMode, algoStr);
```
Comment (HendrikLeuschner) [inline openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java]: Is it correct to enforce the time dependent weighting to be an access weighting here? should it not be something more general?

File: openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphStorageFactory.java [hunk 3]
```diff
@@ -86,9 +86,6 @@ else if (geTurnCosts != null && !graphExtensions.isEmpty()) {
 				((ORSGraphHopper) gh).initCoreLMAlgoFactoryDecorator();
 		}
 
-		if (gh.getLMFactoryDecorator().isEnabled())
```
Comment (HendrikLeuschner) [inline openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphStorageFactory.java:91]: Why was this removed?

Comment (aoles) [reply openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphStorageFactory.java:91]: Fair point, this is quite obscure. This change was complemented by https://github.com/GIScience/graphhopper/commit/b44430e3ded517b4f161bd5ff16e7c495c04a1a1 and the rearrangement was necessary in order to get things to work. Unfortunately, I don't remember the exact details anymore, but if you follow the program flow then you will realize that before this change we had to call `initLMAlgoFactoryDecorator();` only because the original call in GH never happened because of this very modification in the fork:

```//ORS-GH MOD START
        if (graphStorageFactory != null) {
            ghStorage = graphStorageFactory.createStorage(dir, this);
        }

        if (ghStorage == null) {
        //ORS-GH MOD END```

Comment (HendrikLeuschner) [reply openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphStorageFactory.java:91]: Okay, seems reasonable to me. The order of initialization is also kept intact. As long as the code in ORSGraphStorageFactory that follow the initialization of core and corelm does not depend on LM, I do not see a reason why we should not do it on the "outer layer". But why is the CH part still in this factory and not in the graphhopper code?

Comment (aoles) [reply openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphStorageFactory.java:91]: The CH part has never been a subject to time-dependent routing and as such has not been modified in these PRs. 

The change to LM creation was necessary because of the fact that currently `TimeDependentFastestWeighting` requires `GraphHopperStorage` to be initialized beforehand in order to create the `SpeedCalculator`. The original code arrangement failed with a `NullPointerException`.

File: openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java (hunk 2 above)
Comment (aoles) [reply openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java]: This call is supposed to wrap any super `weighting` (say `shortest`, `fastest`, or `td_fastest` into a `TimeDependentAccessWeighting`  similarly as `createTurnWeighting`, so I don't really see an issue here.

Comment (HendrikLeuschner) [reply openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java]: I understand that, I was wondering whether this would be correct under any circumstance, or whether a generic td weighting would be better. This generic weighting can then be made an accessWeighting if the conditions (I do not know whats necessary for that) are met

File: openrouteservice/src/main/java/org/heigit/ors/routing/RoutingProfile.java (hunk 1 above)
Comment (aoles) [reply openrouteservice/src/main/java/org/heigit/ors/routing/RoutingProfile.java:1009]: Thanks for these suggestions! The logic was adapted from the original tardur implementation which indeed has the td routing enabled just for car and bike encoders. I can look into improving this part.

File: openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java (hunk 2 above)
Comment (aoles) [reply openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java]: So as long as time-dependent routing is enabled and a conditional access storage is available to a given encoder it makes sense to have `TimeDependentAccessWeighting` enabled for the sake of correctness of routes. Do you have any particular time-dependent weighting in mind which would fit into this scheme? When it comes to the fastest-based weightings: it doesn't seem feasible to me to have them implemented as wrapper weightings because of the conversion from speed to the time domain which happens in them internally.

Comment (HendrikLeuschner) [reply openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java]: I do not have one in mind. But we might come to that point later. Is the AccessStorage always available? Are there circumstances where it is not created?

File: openrouteservice/src/main/java/org/heigit/ors/routing/RoutingProfile.java (hunk 1 above)
Comment (aoles) [reply openrouteservice/src/main/java/org/heigit/ors/routing/RoutingProfile.java:1009]: I've made some adaptations. Actually `td_fastest` has been implemented and tested just with the car profile, so I have kept that logic. This part will be updated anyway once we introduce proper td routing.

File: openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java (hunk 2 above)
Comment (aoles) [reply openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java:364]: There is currently no mechanism implemented which would prevent it from being created.

Comment (aoles) [reply openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java]: OK, seems that it would be reasonable to add to `createTimeDependentAccessWeighting` method a check for the presence of `conditional_access` encoded value for a given flag encoder. I will fix this.

Comment (aoles) [reply openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/ORSGraphHopper.java]: Done!

**Review by HendrikLeuschner — APPROVED** [review]
Looks fine to me now.
